### ISSUE TITLE
spack load does not load external modules

### ISSUE THREAD

**Issue by koysean** [LLNL/spack#35168]
### Steps to reproduce

Install an external package specified with modules and load it, e.g.:

```console
$ module purge
$ spack install mpich
==> mpich@50.1 : has external module in [MODULE_LIST]
[+] /path/to/mpich (external mpich-50.1-yuxbwgffkpgydwdllhfoyrrk35x52h36)
$ spack load mpich
$ spack load --list
==> 1 loaded package
-- linux-sles15-x86_64 / oneapi@ --------------
mpich@50.1
$ module list
No modules loaded
```

Spack loads mpich internally but does not load the corresponding modules in MODULE_LIST. This is an issue when the external modules contain environment modifications beyond just setting the typical paths from a prefix inspection.

Expected behavior would be to load the modules directly or otherwise produce the equivalent environment modifications (read from the modulefiles). Currently, the environment modifications are taken from a prefix inspection of `/path/to/mpich`, then from the dependencies (there aren't dependencies in this example), and finally from the `setup_run_environment()` method. Even when mpich is loaded into the environment as a dependency (e.g. with `spack install hypre^mpich; spack load hypre`), no modules are loaded.

### Information on your system

* **Spack:** 0.20.0.dev0 (5f8c09fd334d729cf2ff0cdd2398ead78d9c621c)
* **Python:** 3.6.15
* **Platform:** linux-sles15-icelake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by koysean** [comment]
Digging deeper into this, it seems like modules are only loaded into the build environments, but not in the user environment at runtime. Additionally, there is a disconnect between the way that Spack handles its own environment modifications (`EnvironmentModifications`) and the way that external modules are handled (with `module()` in `spack.util.module_cmd`). It seems to me that the simplest solution here is to use `EnvironmentModifications` to handle external modules as well. This allows for easier integration with the current implementations of `spack load` and `spack unload`, which both use `EnvironmentModifications`. Does this sound like a reasonable solution? I can try implementing it myself but I don't want to commit to it before getting input from a more experienced dev.

**Comment by shahzebsiddiqui** [comment]
to see your loaded specs you need to do `spack load --list`

**Comment by koysean** [comment]
Thanks. The issue is not that spack isn't loading mpich internally but that spack doesn't look at the mpich modulefile(s) when `spack load` is called. I edited the original post to add the `spack load --list` output and clarify a bit about why this is a problem.

**Comment by shahzebsiddiqui** [comment]
Yeah so if you are expecting `spack load mpich` to reflect changes in `module list` that is not going to happen. `spack load`  works independent of `module load` and you can just check what gets updated in your environment by using `spack load --sh` it will print a change of all environment, PATH that will be updated.

The way you can test if your environment is loading properly is to manually test the compiler or build some code

```
spack load  mpich
mpicc --version
mpicc hello.c -o hello
mpirun -np 2 ./hello
```

The `mpicc` or `mpirun` can be swapped but you get the idea.

**Comment by giordano** [comment]
The problem I mentioned the other day on Slack is that on a system I have access to cp2k is a system module, but executables are installed in a non-standard place (i.e., not in `${prefix}/bin`), so that adding system cp2k as external package to a Spack environment is useless, because spack will never be able to find the executables, not even if specifying the `module` (which does set `PATH` correctly) instead of the prefix.

**Comment by koysean** [comment]
> Yeah so if you are expecting `spack load mpich` to reflect changes in `module list` that is not going to happen.

This is not what I meant necessarily. What @giordano and I are expecting is that `spack load` on an external package should produce the same environment modifications as those provided by its modulefile, independent of whether or not lmod/tcl are aware of it.

So in my example, there shouldn't be a difference between running `module load mpich` and `spack load mpich` as far as environment variables like PATH, LD_LIBRARY_PATH, etc. are concerned.

To reiterate, these environment modifications do occur at build time, just not when `spack load` is invoked at runtime. Unfortunately the current implementation of the modulefile handling (at buildtime) involves environment changes that are not reversible - this is why I suggested utilizing the `EnvironmentModifications` interface to handle modulefile environment modifications.
